Document deploying ASO in multitenant mode (#2019)
Azure/azure-service-operator: docs/hugo/content/introduction/multitenant-deployment.md

# Deploying ASO v2 in multi-tenant mode

The current release version of the ASO deployment YAML installs the operator in single-tenant mode:

> it's deployed in the `azureserviceoperator-system` namespace with one set of Azure credentials, and manages resources in any namespace in the cluster.

Reviewer: Should this be like a bulleted list or something? It almost reads like one.

Author: Do you mean something like this?

> The current release version of the ASO deployment YAML installs the operator in single-tenant mode. It:
> * is deployed in the `azureserviceoperator-system` namespace,
> * has a single set of Azure credentials,
> * manages Azure resources in any namespace,
> * handles webhooks fired when any Azure resource is changed.

That feels a bit clunky to me, but if you think it's clearer I'll update it.

Reviewer: I did mean that, but agree it is a bit clunky so maybe just leave it as is.

That single operator deployment handles webhooks fired when Azure resources are changed.

ASO can also be deployed in a multi-tenant configuration that enables using separate credentials to manage resources in different Kubernetes namespaces.

Running the operator in multi-tenant mode requires one deployment to handle webhooks (since webhook configurations are cluster-level resources) and then a deployment for each tenant, each with its own credentials and set of namespaces that it watches for Azure resources.
To deploy the operator in multi-tenant mode the release YAML has been split into two parts:

1. Cluster-wide resources:

>    * Custom resource definitions for the Azure resources

Reviewer: minor: consistency - you end some of these with a period and some without. I think they should all end in `.`

   * Cluster roles for managing those resources

>    * The `azureserviceoperator-system` namespace containing the deployment and service to run the operator.

Reviewer: A bit confused about this. In multitenant mode this namespace would just be for the webhook deployment, since it's running in webhook only mode, right?

Author: Clarified this (in conjunction with the comment below).

>      The webhook deployment of the operator is configured to run in webhook-only mode.

Reviewer: ```suggestion
     The webhook deployment if the operator is configured to run in webhook-only mode.
```

I think?

Reviewer: Combine with the above comment though I think it might be clearer to just talk about this as the webhooks for the multi-tenant operator (or something) though, rather than calling it "the operator" which confused me a bit because there's kinda more than one of them.

Author: Yeah, I agree - I'll reword it to make that clearer.

   * The `aso-controller-settings` secret defining the Azure credentials that should be used, and the set of namespaces that this operator will watch for Azure resources.
   * A cluster role binding enabling the per-tenant operator's service account to manage the Azure resources.

## Example files
Examples of the deployment YAML files are available on the release page for ASO v2 releases from [v2.0.0-alpha.5](https://github.com/Azure/azure-service-operator/releases/tag/v2.0.0-alpha.5).
The cluster-wide file `multitenant-cluster_v2.0.0-alpha.5.yaml` can be used as-is (the webhook operator namespace is fixed as `azureserviceoperator-system`),
but the namespaces and cluster role binding in the per-tenant file `multitenant-tenant_v2.0.0-alpha.5.yaml` will need to be updated from `tenant1` to the desired name for each tenant.

## Per-tenant configuration
Create the `aso-controller-settings` secret as described in the [authentication docs](https://azure.github.io/azure-service-operator/introduction/authentication/),
but create the secret in the tenant namespace and add an extra target namespaces key to it:
```

> export TENANT_NAMESPACE="<tenant namespace>"

Reviewer: Nice, this is very clear.

export AZURE_SUBSCRIPTION_ID="<subscription id>"
export AZURE_TENANT_ID="<tenant id>"
export AZURE_CLIENT_ID="<client id>"
export AZURE_CLIENT_SECRET="<client secret>"
export AZURE_TARGET_NAMESPACES="<comma-separated namespace names>"
kubectl create namespace "$TENANT_NAMESPACE"
cat <<EOF | kubectl apply -f -
apiVersion: v1
kind: Secret
metadata:
  name: aso-controller-settings
  namespace: $TENANT_NAMESPACE
stringData:
  AZURE_SUBSCRIPTION_ID: "$AZURE_SUBSCRIPTION_ID"
  AZURE_TENANT_ID: "$AZURE_TENANT_ID"
  AZURE_CLIENT_ID: "$AZURE_CLIENT_ID"
  AZURE_CLIENT_SECRET: "$AZURE_CLIENT_SECRET"
  AZURE_TARGET_NAMESPACES: "$AZURE_TARGET_NAMESPACES"
EOF
```

## Role handling

> The multi-tenant deployment example files have a cluster role granting access to the Azure resource types whichever namespaces they are created in,

Reviewer: Phrasing? This part reads weird to me:
"have a cluster role granting access to the Azure resource types whichever namespaces they are created in"

Author: Yeah, reads like a partial edit... 
Ok, I've updated the wording for that section.

and then a binding to that cluster role for the service account in each tenant-operator namespace.
Each cluster role binding is named for the specific tenant so they don't collide and can be managed separately.
This is convenient since it means that you don't need to permit access for Azure resources in each of the target namespaces individually,
but for some usage scenarios it might be too permissive.

In those cases the `azureserviceoperator-manager-role` should be changed from a `ClusterRole` into `Role`s in each of the target namespaces (where the Azure resources will be created, rather than where the operator pods run),
and a `RoleBinding` should be created in that namespace linking the `Role` to the service account for the tenant operator that will be managing Azure resources in this target namespace.

## Updgrading
When upgrading to a newer version of ASO the cluster-wide resources (CRDs, cluster roles) and the webhook operator deployment must be upgraded before upgrading the tenant operators.

> They can be upgraded by applying the new YAML for that type.

Reviewer: Is there an ordering requirement between the CRDs and the webhooks, or the CRDs and the cluster roles?

Author: Yes, it should be CRDs then (webhooks and cluster roles). I'll clarify that - what I really mean is that you'd apply the upgraded cluster YAML file first (which will have the changes in the right order), and then the upgraded tenant YAML files.

Author: Ha, "Updgrading"

# Deploying ASO v2 in multi-tenant mode

> The current release version of the ASO deployment YAML installs the operator in single-tenant mode:
> it's deployed in the `azureserviceoperator-system` namespace with one set of Azure credentials, and manages resources in any namespace in the cluster.

Reviewer 2: Perhaps tweak the wording/formatting to emphasize what's different about single tenant mode?
```suggestion
The current release version of the ASO deployment YAML installs the operator in _single-tenant_ mode:
deployed in the `azureserviceoperator-system` namespace with a _single set of Azure credentials_, and manages resources in _any namespace_ in the cluster.
```

That single operator deployment handles webhooks fired when Azure resources are changed.

> ASO can also be deployed in a multi-tenant configuration that enables using separate credentials to manage resources in different Kubernetes namespaces.

Reviewer 2: ```suggestion
ASO may also be deployed in a _multi-tenant_ configuration, enabling separate credentials for managing resources in different Kubernetes namespaces.
```

> Running the operator in multi-tenant mode requires one deployment to handle webhooks (since webhook configurations are cluster-level resources) and then a deployment for each tenant, each with its own credentials and set of namespaces that it watches for Azure resources.
> To deploy the operator in multi-tenant mode the release YAML has been split into two parts:

Reviewer 2: Minor suggestions for clarity.

```suggestion
Running the operator in multi-tenant mode uses one deployment to handle webhooks (required because webhook configurations are cluster-level resources), and then a separate deployment for each tenant, each with its own credentials and a set of namespaces that it watches for Azure resources.

To deploy the operator in multi-tenant mode the release YAML has been split into two parts:
```

1. **Cluster-wide resources**:
   * Custom resource definitions for the Azure resources.

   * Cluster roles for managing those resources.

>    * The `azureserviceoperator-system` namespace containing the deployment and service to handle ASO webhooks.
>      The webhook service is a deployment of the ASO operator image, but configured to run in webhook-only mode.
>      It won't try to reconcile Azure resources with ARM, and so doesn't need Azure credentials.

Reviewer 2: Any acronym should be expanded in place and the sentence should make sense.

```suggestion
   * The `azureserviceoperator-system` namespace containing the deployment and service to handle ASO webhooks.
     The webhook service is a deployment of ASO, but configured to run in _webhook-only_ mode.
     It will not try to reconcile Azure resources with ARM, and so does not need any Azure credentials.
```

Author: Good call about the acronym, I've added that early in the doc. I'm keeping the distinction that the webhook service is a deployment of the ASO image based on a comment from @matthchr - he felt that calling the webhook service a deployment of the operator was a bit confusing and I agree.

I'm keeping the contractions because I prefer them! 😝

   * Webhook configuration referring to that service.

2. **Per-tenant resources**:
   * A namespace containing the deployment to run the tenant operator, configured for watchers-only mode.

   * The `aso-controller-settings` secret defining the Azure credentials that should be used, and the set of namespaces that this operator will watch for Azure resources.

   * A cluster role binding enabling the per-tenant operator's service account to manage the Azure resources.

## Example files
Examples of the deployment YAML files are available on the release page for ASO v2 releases from [v2.0.0-alpha.5](https://github.com/Azure/azure-service-operator/releases/tag/v2.0.0-alpha.5).

> The cluster-wide file `multitenant-cluster_v2.0.0-alpha.5.yaml` can be used as-is (the webhook deployment namespace is fixed as `azureserviceoperator-system`),
> but the namespaces and cluster role binding in the per-tenant file `multitenant-tenant_v2.0.0-alpha.5.yaml` will need to be updated from `tenant1` to the desired name for each tenant.

Reviewer 2: Trying to make is more apparent that a separate file is needed for each tenant.

```suggestion
The cluster-wide file `multitenant-cluster_v2.0.0-alpha.5.yaml` can be used as-is (the webhook deployment namespace is fixed as `azureserviceoperator-system`),
but the namespaces and cluster role binding in the per-tenant file `multitenant-tenant_v2.0.0-alpha.5.yaml` will need to be customized in each tenant's configuration file from `tenant1` to the desired name for that tenant.
```

Author: Except I used `YAML file` rather than `configuration file` because I don't want it to be confused with the configuration in the secret below.

## Per-tenant configuration
Create the `aso-controller-settings` secret as described in the [authentication docs](https://azure.github.io/azure-service-operator/introduction/authentication/),

> but create the secret in the tenant namespace and add an extra target namespaces key to it:

Reviewer 2: Maybe give an example of what a tenant namespace would look like?

Author: Do you mean the contents of the tenant namespace after deployment? Or an example of a tenant name?
I'm assuming the former.

Reviewer 2: I actually meant the later, but showing the former would be good too! I think the diagrams will help here when you embed them inline.

export AZURE_CLIENT_SECRET="<client secret>"
export AZURE_TARGET_NAMESPACES="<comma-separated namespace names>"
kubectl create namespace "$TENANT_NAMESPACE"
cat <<EOF | kubectl apply -f -
apiVersion: v1
kind: Secret
metadata:
  name: aso-controller-settings
  namespace: $TENANT_NAMESPACE
stringData:
  AZURE_SUBSCRIPTION_ID: "$AZURE_SUBSCRIPTION_ID"
  AZURE_TENANT_ID: "$AZURE_TENANT_ID"
  AZURE_CLIENT_ID: "$AZURE_CLIENT_ID"
  AZURE_CLIENT_SECRET: "$AZURE_CLIENT_SECRET"
  AZURE_TARGET_NAMESPACES: "$AZURE_TARGET_NAMESPACES"
EOF
```

## Role handling
The multi-tenant deployment example files have a single `ClusterRole` that grants access to the Azure resource types,
and then a binding to that `ClusterRole` for the service account in each tenant-operator namespace.
Each `ClusterRoleBinding` is named for the specific tenant so they don't collide and can be managed separately.

This is convenient since there's no need to permit access for Azure resources in each of the target namespaces individually,
but it means that the only thing preventing one tenant operator from reading another's resources is the `AZURE_TARGET_NAMESPACES` setting for each operator.

> For some usage scenarios that might be too permissive.

Reviewer 2: Blank line before, so the line stands out more to someone reading the file.

```suggestion

For some usage scenarios that might be too permissive.
```

In those cases the `azureserviceoperator-manager-role` should be changed from a `ClusterRole` into `Role`s in each of the target namespaces (where the Azure resources will be created, rather than where the tenant-operator pods run),

> and a `RoleBinding` should be created in that namespace linking the `Role` to the service account for the tenant operator that will be managing Azure resources in this target namespace.

Reviewer 2: Would this be clearer with a couple of diagrams showing the differences between scenarios?

Author: Yeah, good idea - I'll have a go at knocking them up in graphviz.
